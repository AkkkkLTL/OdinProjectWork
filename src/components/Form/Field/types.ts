export interface IFieldProps {
  name:string;
  label:string;
  type?: "Text" | "Email" | "Number" | "Select" | "TextArea";
  options?:string[]
}