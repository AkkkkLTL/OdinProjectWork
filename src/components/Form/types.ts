import { ReactNode } from "react";

export interface IValues {
  [key:string]: any;
}

export type Validator = (fieldName:string, values:IValues, args?:any) => string;

export interface IFormProps {
  defaultValues: IValues;
  validationRule: IValidationProp;
  onSubmit: (values:IValues) => Promise<ISubmitResult>;
  children: ReactNode;
}
interface IState {
  values: IValues;
  errors: IErrors;
  submitting: boolean;
  submitted: boolean;
}

export interface IFormContext {
  errors: IErrors;
  values: IValues;
  setValue?: (fieldName:string, value:any) => void;
  validate?:(fieldName:string, value:any) => void;
}

interface IValidation {
  validator: Validator;
  arg?:any;
}

interface IValidationProp {
  [key: string]: IValidation | IValidation[];
}

export interface IErrors {
  [key:string]: string[];
}

export interface ISubmitResult {
  success: boolean;
  errors?: IErrors;
}