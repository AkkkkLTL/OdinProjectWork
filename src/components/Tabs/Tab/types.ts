import { ReactNode } from "react";

export interface ITabProps {
  name: string;
  initialActive?: boolean;
  children?: ReactNode;
  heading: () => string | ReactNode;
}