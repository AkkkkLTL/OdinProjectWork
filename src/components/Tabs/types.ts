import { ReactNode } from "react";

export interface ITabsProps {
  children?: ReactNode;
}

export interface ITabsContext {
  activeName?: string;
  handleTabClick?: (name:string, content:ReactNode) => void;
}
