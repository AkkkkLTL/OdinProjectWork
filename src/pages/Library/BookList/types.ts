import { Author, Book } from "@router/Library/loaders/types"

export type BookListLoader = {
  books: Book[],
  authors: Author[],
}

export type FieldType = {
  title:string,
  isbn:string,
  author:string,
  pages?:number,
  currentPages?:number,
  cover:string,
  status?:string,
} & Book;