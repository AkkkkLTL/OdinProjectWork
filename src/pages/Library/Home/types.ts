import { Book, SummaryData } from "@/router/Library/loaders/types"

export type homeLoader = {
  summaryData: SummaryData[],
  readingBooks: Book[],
}

export type dataType = SummaryData[];
export type optionType = {
  value: number,
  name: string,
  itemStyle: {
    color: string
  },
}

export type FileType = {
  pages: number,
  currentPages: number,
}