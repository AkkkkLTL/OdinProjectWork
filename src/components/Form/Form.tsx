import { FC } from "react";
import { IFormProps } from "./types";
import useForm from "./useForm";
import { FormContext } from "./contex";

const Form:FC<IFormProps> = (props) => {

  const {
    submitting, submitted,
    context,
    handleSubmit
  } = useForm(props);

  return (
    <FormContext.Provider value={context}>
      <form
        className="form"
        noValidate={true}
        onSubmit={handleSubmit}
      >
        <div>
          {props.children}
        </div>
        <div className="form-group">
            <button 
              type="submit"
              disabled={submitting || submitted}
            >
              Submit
            </button>
          </div>
      </form>
    </FormContext.Provider>
  )
}

export default Form;