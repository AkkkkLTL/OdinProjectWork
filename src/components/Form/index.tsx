import { default as ContainerForm } from "./Form";
import Field from "./Field";
import "./styles.scss";

type CompondComponent = typeof ContainerForm & {
  Field: typeof Field
};

const Form = ContainerForm as CompondComponent;

Form.Field = Field;

export default Form;