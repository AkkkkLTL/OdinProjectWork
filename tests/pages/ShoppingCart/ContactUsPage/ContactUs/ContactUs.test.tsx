import { afterEach, describe, expect, test, vitest } from "vitest";
import { cleanup, fireEvent, render } from "@testing-library/react";

import ContactUs from "@/pages/ShoppingCart/ContactUsPage/ContactUs"
import { ISubmitResult } from "@/components/Form/types";

describe("ContactUs", () => {
  afterEach(cleanup)

  test("When submit without filling in fields should display errors", () => {
    const handleSubmit = vitest.fn();
    const { getAllByText, getByText } = render(<ContactUs onSubmit={handleSubmit} />);
    
    const submitButton = getByText("Submit");
    fireEvent.click(submitButton);

    const errorSpans = getAllByText("This must be populated");
    expect(errorSpans.length).toBe(2);
  });

  test("When submit after filling in fields should sumit okay", () => {
    const handleSubmit = vitest.fn();

    const { container, getByText, getByLabelText } = render(<ContactUs onSubmit={handleSubmit} />);
    const nameField:HTMLInputElement = getByLabelText("Your Name") as HTMLInputElement;
    expect(nameField).not.toBeNull();
    fireEvent.change(nameField, {target: {value: "Carl"}});

    const emailField:HTMLInputElement = getByLabelText("Your Email Address") as HTMLInputElement;
    expect(emailField).not.toBeNull();
    fireEvent.change(emailField, {target: {value: "carl.rippon@testmail.com"}});

    const submitButton = getByText("Submit");
    fireEvent.click(submitButton);
    const errorsDiv = container.querySelector("[data-testid='formErrors'");
    expect(errorsDiv).toBeNull();
    expect(handleSubmit).toBeCalledTimes(1);
    expect(handleSubmit).toBeCalledWith({
      name: "Carl",
      email: "carl.rippon@testmail.com",
      reason: "Support",
      notes: ""
    });
  });

  test("Renders okay", () => {
    const handleSubmit = async ():Promise<ISubmitResult> => {
      return {
        success: true
      }
    };

    const { container } = render(<ContactUs onSubmit={handleSubmit} />);
    expect(container).toMatchSnapshot();
  });
});