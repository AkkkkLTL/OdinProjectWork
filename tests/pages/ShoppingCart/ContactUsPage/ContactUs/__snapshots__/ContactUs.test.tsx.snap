// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ContactUs > Renders okay 1`] = `
<div>
  <form
    class="form"
    novalidate=""
  >
    <div>
      <div
        class="form-group"
      >
        <label
          for="name"
        >
          Your Name
        </label>
        <input
          id="name"
          type="text"
          value=""
        />
      </div>
      <div
        class="form-group"
      >
        <label
          for="email"
        >
          Your Email Address
        </label>
        <input
          id="email"
          type="text"
          value=""
        />
      </div>
      <div
        class="form-group"
      >
        <label
          for="urgency"
        >
          How Urgent Is a Response
        </label>
        <input
          id="urgency"
          type="number"
          value=""
        />
      </div>
      <div
        class="form-group"
      >
        <label
          for="reason"
        >
          Reason You Need to Contact Us
        </label>
        <select>
          <option
            value="Marking"
          >
            Marking
          </option>
          <option
            value="Support"
          >
            Support
          </option>
          <option
            value="Feedback"
          >
            Feedback
          </option>
          <option
            value="Jobs"
          >
            Jobs
          </option>
          <option
            value="Other"
          >
            Other
          </option>
        </select>
      </div>
      <div
        class="form-group"
      >
        <label
          for="notes"
        >
          Additional Notes
        </label>
        <textarea
          id="notes"
        />
      </div>
    </div>
    <div
      class="form-group"
    >
      <button
        type="submit"
      >
        Submit
      </button>
    </div>
  </form>
</div>
`;
